VMware: fix customization.hostname handling

dgmorales opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Stops silently mangling of the user supplied customization.hostname to fit VMware API requirements (unless --force is set). Instead it throws a helpful error message if the name does not fit the requirements.

This PR is an alternative proposal to https://github.com/ansible/ansible/pull/46056.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
lib/ansible/modules/cloud/vmware/vmware_guest.py

##### ADDITIONAL INFORMATION

This is related to a discussion that started in https://github.com/ansible/ansible/pull/38005#issuecomment-420568723 and was followed up on PR https://github.com/ansible/ansible/pull/46056.

In the code proposed here:

* customization.hostname is not required, get it's default from vm name
* if force==False, no silently hostname mangling is done. If name is not valid, a helpful message is throw
* Also checks that the name contains at least one letter, as said in the [vmware doc](https://pubs.vmware.com/vsphere-51/index.jsp?topic=%2Fcom.vmware.vsphere.vm_admin.doc%2FGUID-9A5093A5-C54F-4502-941B-3F9C0F573A39.html).
* if force==True, silently mangle the name or customization.hostname as currently done (unless name has no letter)

I didn't test it against a VMware enviroment yet. I will do it soon on the following days. But I did try to test the logic mocking an object with a params attribute. Here's the results of my testing:

_(first hostname is **vm name**, followed by FAILED or final hostname)_

With customization.hostname unset and force==False:
```text
hostname "vmware01.example.com" ... FAILED
hostname "host.d" ... FAILED
hostname "a xyz" ... FAILED
hostname "90921" ... FAILED
hostname "-0" ... FAILED
hostname "a" ... a
hostname "A" ... A
hostname "xyz-aa-0" ... xyz-aa-0
hostname "2a131" ... 2a131
hostname "vmware01" ... vmware01
hostname "Vmware" ... Vmware
```

With customization.hostname unset and force==True
```text
hostname "vmware01.example.com" ... vmware01
hostname "host.d" ... host
hostname "a xyz" ... axyz
hostname "90921" ... FAILED
hostname "-0" ... FAILED
hostname "a" ... a
hostname "A" ... A
hostname "xyz-aa-0" ... xyz-aa-0
hostname "2a131" ... 2a131
hostname "vmware01" ... vmware01
hostname "Vmware" ... Vmware
```

Also: 
- Setting customization.hostname to "vmware-01.example.com" and Force==False fails every case
- Setting customization.hostname to "vmware-01.example.com" and Force==True returns "vmware-01" to every case

dgmorales committed:
VMware: fix customization.hostname handling

- Stops silently mangling of the user supplied customization.hostname to
  fit VMware API requirements (unless --force is set). Instead it throws a
  helpful error message if the name does not fit the requirements.
- Fix existing test case with VM name that does not fit the
  requirements.
- Add a new test case for customization hostname validation

dgmorales commented:
cc @Akasurde @vincenzodnp @pdellaert

dgmorales commented:
I see there some errors detected by the linter, will work on that.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -295,8 +295,9 @@
-    - ' - C(hostname) (string): Computer hostname (default: shorted C(name) parameter). Allowed characters are alphanumeric (uppercase and lowercase)
-          and minus, rest of the characters are dropped as per RFC 952.'
+    - ' - C(hostname) (string): Computer hostname (default: C(name) parameter). Allowed characters are alphanumeric (uppercase and lowercase)
+          and minus (no dots allowed, VMware guest customization for linux does not support receiving a FQDN name). Use "force=True" for old
```
```suggestion
          and minus (no dots allowed, VMware guest customization for Linux does not support receiving a FQDN name). Use "force=True" for old
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1552,10 +1553,21 @@ def customize_vm(self, vm_obj):
+                    # but name must have at least one letter
+                    hostname = re.sub(r"[^a-zA-Z0-9\-]", "", hostname.split('.')[0])
+                else:
+                    self.module.fail_json(msg="The hostname you have set it not valid for use with vmware guest customization for linux."
```
```suggestion
                    self.module.fail_json(msg="The hostname you have set it not valid for use with vmware guest customization for Linux."
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -295,8 +295,9 @@
-          and minus, rest of the characters are dropped as per RFC 952.'
+    - ' - C(hostname) (string): Computer hostname (default: C(name) parameter). Allowed characters are alphanumeric (uppercase and lowercase)
+          and minus (no dots allowed, VMware guest customization for linux does not support receiving a FQDN name). Use "force=True" for old
+          pre 2.8 behavior of automatically fixing the hostname by removing the invalid chars and domain name.
```
```suggestion
          pre 2.8 behavior of automatically fixing the hostname by removing the invalid chars and domain name.'
```

Akasurde replied:
Document linter is failing due to missing single quote at end of this line. :)

Akasurde commented:
@pdellaert @dericcrago @Im0 @ckotte @jeking3 Could you please review this ? Thanks in advance.

Akasurde commented:
Rest LGTM

mattclay commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -295,8 +295,9 @@
-    - ' - C(hostname) (string): Computer hostname (default: shorted C(name) parameter). Allowed characters are alphanumeric (uppercase and lowercase)
-          and minus, rest of the characters are dropped as per RFC 952.'
+    - ' - C(hostname) (string): Computer hostname (default: C(name) parameter). Allowed characters are alphanumeric (uppercase and lowercase)
+          and minus (no dots allowed, VMware guest customization for linux does not support receiving a FQDN name). Use "force=True" for old
```
Making the new behavior the default is not backwards compatible. The new behavior should be opt-in rather than opt-out to avoid causing problems with existing playbooks during an upgrade.

dgmorales replied:
An important general rule indeed, but IMHO this should be an exception. 

Because I believe the current behaviour is misleading. Not sure if it's the best word to describe, but It made me waste time figuring out why the domain name was not being configured in the OS hostname as I had instructed the module to do. Instead it could just have thrown an error telling me it wasn't supported. Can happen to many others, probably have already.

I also believe that the likeliness of someone out there relying on this behaviour is not that big. It could surely be masking an configuration error that will be revealed with this, but should be an easy fix, and people should validate/test new ansible versions before adopting then anyway.

Could you give use some advice on that @Akasurde ?

If we make the new behaviour opt-in, then it would need to be something more than just trigger the new behaviour  with --force, because that would be also confusing. Any suggestions in that case?

jeking3 replied:
@mattclay:
Status
This module is not guaranteed to have a backwards compatible interface. [preview]

It is acceptable by this definition to change the behavior, document said change with a changelog fragment as a breaking change.  This of course will restrict it to Ansible 2.9 as the next delivery vehicle.

Normally I prefer to see changes that are backwards compatible, however in this case Ansible's own status indicates it is not required, and "my.invalid.hostname" is not a valid hostname.

mattclay replied:
Be sure to include a changlog entry explaining the backwards compatibility issue.

dgmorales commented:
Well, now it's throwing an error about the hostname given in one of the tests. I'll check it out and propose a fix for the test, if that's the case.

When done should I squash my commits and push force, or leave the squashing for merge time?

dgmorales commented:
I have:
- Fixed a test case that used a name that didn't fit there requirements of the VMware API
- Added a test case for customisation hostname validation
- Squashed my commits and rebased to current devel
- Fixed another syntax error detected by the CI

I've ran some Ansible-test sanity test locally and it went. I'm waiting on shippable CI, hope it passes now.

dgmorales commented:
While I was doing that I've began to question if the same approach should be applied to the windows customisation code, since this is just for linux, but the docs says the requirements are the same (see this [doc](https://docs.vmware.com/en/VMware-vSphere/6.5/com.vmware.vsphere.vm_admin.doc/GUID-CAEB6A70-D1CF-446E-BC64-EC42CDB47117.html)). The windows code is very different and I'm not sure if it's a good idea to consider changing that in this same PR.

goneri commented on test/integration/targets/vmware_guest/tasks/customization_hostname_validation.yml:
```diff
@@ -0,0 +1,121 @@
+# Copyright: (c) 2019, Diego Morales <dgmorales@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- name: Wait for Flask controller to come up online
```
Hi @dgmorales dgmorales,

We are refactoring the vmware test-suite to keep it simple and be able to run it on a regular VMware environment.
The test roles should not interact with vcsim themselves anymore. You just need instead to import the `prepare_vmware_tests` role with the correct parameters. In addition, it's also important that you clean the resources that you create.

Example: https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_guest_find/tasks/main.yml
